src/java/org/apache/cassandra/utils/btree/BTree.java

```diff
@@ -20,283 +20,547 @@
+        if (!isSimple(updateF))
+        {
+            updateF.onAllocated(ObjectSizes.sizeOfReferenceArray(values.length));
+            for (int i = 0; i < size; i++)
+                values[i] = updateF.apply((I) values[i]);
+        }
+        return values;
+    }
 
-        int idx = startIdx;
-        for (int i = 0; i < size; i++)
+    /**
+     * Build a leaf with {@code size} elements taken in bulk from {@code insert}, and apply {@code updateF} to these elements
+     * Do not invoke {@code updateF.onAllocated}.  Used by {@link #buildMaximallyDense} and {@link #buildPerfect} which
+     * track the size for the entire tree they build in order to save on work.
+     */
+    private static <C, I extends C, O extends C> Object[] buildLeafWithoutSizeTracking(BulkIterator<I> insert, int size, UpdateFunction<I, O> updateF)
+    {
+        Object[] values = new Object[size | 1]; // ensure that we have an odd-length array
+        insert.fetch(values, 0, size);
+        if (!isSimple(updateF))
         {
-            K k = iterFunc.nextAt(source, idx);
-            values[i] = updateF.apply(k);
-            idx++;
+            for (int i = 0; i < size; i++)
+                values[i] = updateF.apply((I) values[i]);
         }
-        if (updateF != UpdateFunction.<K>noOp())
-            updateF.allocated(ObjectSizes.sizeOfArray(values));
         return values;
     }
 
-    private static <C, K extends C, V extends C, S> Object[] buildInternal(S source, IteratingFunction<S> iterFunc, int size, int level, int startIdx, UpdateFunction<K, V> updateF)
+    /**
+     * Build a root node from the first {@code size} elements from {@code source}, applying {@code updateF} to those elements.
+     * A root node is permitted to have as few as two children, if a branch (i.e. if {@code size > MAX_SIZE}.
+     */
+    private static <C, I extends C, O extends C> Object[] buildRoot(BulkIterator<I> source, int size, UpdateFunction<I, O> updateF)
     {
-        assert size > 0;
-        assert level >= 0;
-        if (level == 0)
-            return buildLeaf(source, iterFunc, size, startIdx, updateF);
+        // first calculate the minimum height needed for this size of tree
+        int height = minHeight(size);
+        assert height > 1;
 
-        // calcuate child num: (size - (childNum - 1)) / maxChildSize <= childNum
-        int childNum = size / (TREE_SIZE[level - 1] + 1) + 1;
+        int denseChildSize = denseSize(height - 1);
+        int childCount = size / (denseChildSize + 1) + 1;
 
-        V[] values = (V[]) new Object[childNum * 2];
-        if (updateF != UpdateFunction.<K>noOp())
-            updateF.allocated(ObjectSizes.sizeOfArray(values));
+        return buildMaximallyDense(source, childCount, size, height, updateF);
+    }
+
+    /**
+     * Build a tree containing some initial quantity of dense nodes, up to a single node of arbitrary size (between
+     * dense and sparse), and the remainder of nodes all sparse.
```

Comment:
The shared `sizeMap` are a pretty minor advantage, given how well their cost amortises. The regularity of the shape permitted more efficient construction, primarily. Also, it's a while back so I may be misremembering, but experimentally I believe this shape also performed better for modifications to the tree than the most obvious alternative of spreading keys evenly through a tree. Since we're copy-on-write this makes sense, as we have to copy any node we modify, and this maximally dense tree has to split fewer nodes (since a preponderance of splits will occur on the dense side of the tree). It has few sparse nodes, but these are more likely to be re-used for splits occurring within their range, and are cheap to copy. But I did not exhaustively examine or consider this, since mostly we construct trees.

Reply:
There are other advantages, I suppose, in that this also standardises the behaviour for inserting and building, with the same conditions being used during overflow, so that functionality is more easily shared (obviously, inserts also utilise the shape of the original tree, but otherwise behave identically to builders).

Of course, sharing the `sizeMap` is still nice, and was a minor beneficial motivation.

Reply:
I would be fine with that change, yes. I agree we are not saving enough with the ‘sizeMap’ here to prefer the current approach, and I doubt it is computationally very different to build versus the approach taken in the builder.

Reply:
Okay, let me prepare a branch that incorporates all comments so far, and that standardizes on the insert layout (if that's not too hard or underperforms), for your to review.

Reply:
👍